overlord/exportstate/handlers_test.go

```diff
@@ -0,0 +1,303 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package exportstate_test
+
+import (
+	"errors"
+	"path/filepath"
+
+	. "gopkg.in/check.v1"
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/overlord"
+	"github.com/snapcore/snapd/overlord/exportstate"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/snaptest"
+	"github.com/snapcore/snapd/testutil"
+)
+
+type handlersSuite struct {
+	testutil.BaseTest
+	o  *overlord.Overlord
+	se *overlord.StateEngine
+	st *state.State
+}
+
+var _ = Suite(&handlersSuite{})
+
+func (s *handlersSuite) SetUpTest(c *C) {
+	s.BaseTest.SetUpTest(c)
+
+	dirs.SetRootDir(c.MkDir())
+	s.AddCleanup(func() { dirs.SetRootDir("") })
+
+	s.o = overlord.Mock()
+	s.se = s.o.StateEngine()
+	s.st = s.o.State()
+	s.AddCleanup(s.se.Stop)
+
+	runner := s.o.TaskRunner()
+	runner.AddHandler("fail", func(task *state.Task, _ *tomb.Tomb) error {
+		return errors.New("injected test failure")
+	}, nil)
+
+	mgr, err := exportstate.Manager(s.st, runner)
+	c.Assert(err, IsNil)
+	s.o.AddManager(mgr)
+	s.o.AddManager(runner)
+	c.Assert(s.o.ExportManager(), Equals, mgr)
+	c.Assert(s.o.StartUp(), IsNil)
+
+}
+
+func (s *handlersSuite) TestDoExportContentNop(c *C) {
+	st := s.st
+	st.Lock()
+	si := &snap.SideInfo{RealName: "snap-name", Revision: snap.R(1)}
+	snaptest.MockSnap(c, "name: snap-name\nversion: 1\n", si)
+	snapstate.Set(st, "snap-name", &snapstate.SnapState{
+		Active:   true,
+		Sequence: []*snap.SideInfo{si},
+		Current:  snap.R(1),
+		SnapType: "app",
+	})
+	change := st.NewChange("change", "...")
+	task := st.NewTask("export-content", "...")
+	task.Set("snap-setup", &snapstate.SnapSetup{
+		SideInfo: &snap.SideInfo{RealName: "snap-name", Revision: snap.R(1)},
+	})
+	change.AddTask(task)
+	st.Unlock()
+
+	c.Check(s.se.Ensure(), IsNil)
+	s.se.Wait()
+
+	st.Lock()
+	defer st.Unlock()
+	c.Check(change.Err(), IsNil)
+	c.Check(task.Status(), Equals, state.DoneStatus)
+	c.Check(change.Status(), Equals, state.DoneStatus)
+}
+
+func (s *handlersSuite) TestDoUnexportContentNop(c *C) {
+	st := s.st
+	st.Lock()
+	si := &snap.SideInfo{RealName: "snap-name", Revision: snap.R(1)}
+	snaptest.MockSnap(c, "name: snap-name\nversion: 1\n", si)
+	snapstate.Set(st, "snap-name", &snapstate.SnapState{
+		Active:   true,
+		Sequence: []*snap.SideInfo{si},
+		Current:  snap.R(1),
+		SnapType: "app",
+	})
+	task := st.NewTask("unexport-content", "...")
+	task.Set("snap-setup", &snapstate.SnapSetup{
+		SideInfo: &snap.SideInfo{RealName: "snap-name", Revision: snap.R(1)},
+	})
+	change := st.NewChange("change", "...")
+	change.AddTask(task)
+	st.Unlock()
+
+	c.Check(s.se.Ensure(), IsNil)
+	s.se.Wait()
+
+	st.Lock()
+	defer st.Unlock()
+	c.Check(change.Err(), IsNil)
+	c.Check(task.Status(), Equals, state.DoneStatus)
+	c.Check(change.Status(), Equals, state.DoneStatus)
+}
+
```

Comment:
we need some sort of idempotence tests as well, about what happens if the task are almost fully rerun (because of a reboot)

Reply:
I will write one but since the primitive used is idempotent and state is not written to disk on success, we should be allrigtht.